src/compiler/checker.ts

```diff
@@ -11635,6 +11661,46 @@ namespace ts {
             return type;
         }
 
+        function instantiationTrampoline(type: Type, mapper: TypeMapper): Type {
+            const genStack = [instantiateStaged(type, mapper)];
+            let result: IteratorResult<[Type, TypeMapper], Type> | undefined;
+            while (genStack.length) {
+                // 300000 stacked up generators is _probably_ going to blow the heap if it continues. Probably.
+                // Downlevel we can only fit a fraction of that - 100000 - so that's what we use here.
+                if (genStack.length > 100000) {
+                    error(currentNode, Diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite);
+                    return errorType;
+                }
+                // Get the top generator off the stack and advance it, passing in nothing if it's said generator's
+                // first invocation, or the result of the prior finished generator otherwise
+                Debug.assert(!result || !isArray(result));
+                result = genStack[genStack.length - 1].next(...(result ? [result.value as Type] : []));
+                // If this generator is done, pop it off the stack and keep around the result for either the parent
+                // generator or the return value
+                if (result.done) {
+                    Debug.assert(result.value && !isArray(result.value));
+                    genStack.pop();
+                    continue;
+                }
+                // Generator is unfinished and returned a type+mapper tuple that it needs resolved - make a generator
+                // for that type+mapper pair and add it to the stack
+                Debug.assert(isArray(result.value));
+                genStack.push(instantiateStaged(...result.value));
+                // Adding a new generator to the stack - consume the result used above so the first
+                // `next` call of the new generator is without arguments
+                result = undefined;
+            }
+            return result!.value as Type;
+        }
+
+        function collectGeneratorReturn<TYield, TStep, TReturn>(gen: Generator<TYield, TReturn, TStep>, feedforward: (yielded: TYield) => TStep): TReturn {

```

Comment:
Seems like the name could just be something like `trampoline`. Do we/could we use this anywhere else? I'd be tempted to say it should go in *core.ts*.
```suggestion
        function trampoline<TYield, TStep, TReturn>(gen: Generator<TYield, TReturn, TStep>, step: (yielded: TYield) => TStep): TReturn {
```

Reply:
`collectGeneratorReturn` isn't the trampoline, though - it's just a naive way to collect the `return` value of a generator (as you may want to do if you're attempting to call one of the inner instantiation functions from outside of the trampoline context (which you shouldn't do because reentrancy is bad but we do anyway)). The base of the `trampoline` is the loop in `instantiationTrampoline`.

Moving it to `core` makes sense, though~